lib/Parse/ParseType.cpp

```diff
@@ -389,64 +389,121 @@ ParserResult<IdentTypeRepr> Parser::parseTypeIdentifier() {
-    EndLoc = consumeStartingGreater();
-  } else {
-    if (Status.isSuccess()) {
-      diagnose(Tok, diag::expected_rangle_protocol);
-      diagnose(LAngleLoc, diag::opening_angle);
-      Status.setIsParseError();
+  // handle deprecated case
+  if (Tok.getKind() == tok::kw_protocol) {
+    SourceLoc ProtocolLoc = consumeToken(tok::kw_protocol);
+    
+    // Check for the starting '<'.
+    if (!startsWithLess(Tok)) {
+      diagnose(Tok, diag::expected_langle_protocol);
+      return nullptr;
+    }
+    SourceLoc LAngleLoc = consumeStartingLess();
+    
+    // Check for empty protocol composition.
+    if (startsWithGreater(Tok)) {
+      SourceLoc RAngleLoc = consumeStartingGreater();
+      auto AnyRange = SourceRange(ProtocolLoc, RAngleLoc);
+      
+      // warn that 'protocol<>' is depreacted and offer to
+      // repalace with the 'Any' keyword
+      diagnose(LAngleLoc, diag::deprecated_any_composition)
+        .fixItReplace(AnyRange, "Any");
+      
+      return makeParserResult(new (Context) ProtocolCompositionTypeRepr(ArrayRef<IdentTypeRepr *>(),
+                                                                        ProtocolLoc,
+                                                                        AnyRange));
+    }
+    
+    // Parse the type-composition-list.
+    ParserStatus Status;
+    SmallVector<IdentTypeRepr *, 4> Protocols;
+    SmallVector<SourceLoc, 4> Commas; // used offer fixit to convert commas to '&'
+    do {
+      // Parse the type-identifier.
+      ParserResult<IdentTypeRepr> Protocol = parseTypeIdentifier();
+      Status |= Protocol;
+      if (Protocol.isNonNull())
+        Protocols.push_back(Protocol.get());
+      Commas.push_back(Tok.getLoc());
+    } while (consumeIf(tok::comma));
+
+    // token after last typename wasn't a comma, don't replace
+    // it with a '&'
+    Commas.pop_back();
+
+    // Check for the terminating '>'.
+    SourceLoc RAngleLoc = PreviousLoc;
+    if (startsWithGreater(Tok)) {
+      RAngleLoc = consumeStartingGreater();
+    } else {
+      if (Status.isSuccess()) {
+        diagnose(Tok, diag::expected_rangle_protocol);
+        diagnose(LAngleLoc, diag::opening_angle);
+        Status.setIsParseError();
+      }
+      
+      // Skip until we hit the '>'.
+      RAngleLoc = skipUntilGreaterInTypeList(/*protocolComposition=*/true);
     }
+    
+    // replace 'protocol<T1, T2>' with 'T1 & T2'
+    auto Diag = diagnose(ProtocolLoc, diag::deprecated_protocol_composition);
+    auto composition = ProtocolCompositionTypeRepr::create(Context, Protocols, ProtocolLoc, {LAngleLoc, RAngleLoc});
+
+    Diag.highlight({ProtocolLoc, RAngleLoc});
+    for (auto Comma : Commas)          // remove commas and '<'
+      Diag.fixItReplace(Comma, " &");
+    Diag
+      .fixItRemove({ProtocolLoc, LAngleLoc}) // remove 'protocol<'
+      .fixItRemove(RAngleLoc); // remove '>'
+
+    return makeParserResult(Status, composition);
+    
+  } else {
```

Comment:
We favor early returns in the compiler (like LLVM), so no need to make this an `else` case.